mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManagerMigration.java

```diff
@@ -0,0 +1,96 @@
+import org.apache.james.user.api.UsersRepository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class CassandraPerUserMaxQuotaManagerMigration implements Migration {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CassandraPerUserMaxQuotaManagerMigration.class);
+    private final UsersRepository usersRepository;
+    private final DomainList domainDao;
+    private final MaxQuotaManager oldMaxQuotaManager;
+    private final MaxQuotaManager newMaxQuotaManager;
+    private UserQuotaRootResolver userQuotaRootResolver;
+
+    @Inject
+    public CassandraPerUserMaxQuotaManagerMigration(UsersRepository usersRepository, DomainList domainDao,
+                                                    @Named("old") MaxQuotaManager oldMaxQuotaManager, MaxQuotaManager newMaxQuotaManager,
+                                                    UserQuotaRootResolver userQuotaRootResolver) {
+        this.usersRepository = usersRepository;
+        this.domainDao = domainDao;
+        this.oldMaxQuotaManager = oldMaxQuotaManager;
+        this.newMaxQuotaManager = newMaxQuotaManager;
+        this.userQuotaRootResolver = userQuotaRootResolver;
+    }
+
+    @Override
+    public void apply() throws InterruptedException {
+        Mono.from(oldMaxQuotaManager.getGlobalMaxMessageReactive()).flatMap(quotaCountLimit -> Mono.from(newMaxQuotaManager.setGlobalMaxMessageReactive(quotaCountLimit))).block();
+        Mono.from(oldMaxQuotaManager.getGlobalMaxStorageReactive()).flatMap(quotaSizeLimit -> Mono.from(newMaxQuotaManager.setGlobalMaxStorageReactive(quotaSizeLimit))).block();
+
+        Flux.fromIterable(getDomains())
+            .flatMap(domain -> Mono.from(oldMaxQuotaManager.getDomainMaxMessageReactive(domain))
+                .flatMap(quotaCountLimit -> Mono.from(newMaxQuotaManager.setDomainMaxMessageReactive(domain, quotaCountLimit)))
+                .then(Mono.from(oldMaxQuotaManager.getDomainMaxStorageReactive(domain))
+                    .flatMap(quotaSizeLimit -> Mono.from(newMaxQuotaManager.setDomainMaxStorageReactive(domain, quotaSizeLimit)))))
+            .then()
+            .block();
+
+        Flux.from(usersRepository.listReactive())
+            .map(userQuotaRootResolver::forUser)
+            .flatMap(quotaRoot -> Mono.from(oldMaxQuotaManager.listMaxMessagesDetailsReactive(quotaRoot))
+                .map(map -> map.get(Quota.Scope.User))
+                .flatMap(quotaCountLimit -> Mono.from(newMaxQuotaManager.setMaxMessageReactive(quotaRoot, quotaCountLimit)))
+                .then(Mono.from(oldMaxQuotaManager.listMaxStorageDetailsReactive(quotaRoot))
+                    .map(map -> map.get(Quota.Scope.User))
+                    .flatMap(quotaSizeLimit -> Mono.from(newMaxQuotaManager.setMaxStorageReactive(quotaRoot, quotaSizeLimit)))))
+            .then()
+            .block();
+    }
```

Comment:
Too much indent means you maybe want to either flatten this or more likely extract a method.